### ISSUE TITLE
[MODULES-65] Deadlock on circular class load 

### ISSUE THREAD

**Issue by dmlloyd** [jboss-modules/jboss-modules#785]
@tdiesler opened **[MODULES-65](https://issues.redhat.com/browse/MODULES-65?redirect=false)** and commented

> git clone -b bootstrap git://github.com/jbosgi/jbosgi-framework.git
> mvn -Dtest=MOD65TestCase install

Running org.jboss.test.osgi.modules.MOD65TestCase
Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.125 sec

---

**Affects:** 1.0.0.Beta13

**Comment by dmlloyd** [comment]
@tdiesler commented

With https://github.com/dmlloyd/jboss-modules/commit/515e04c3710086f9bcf84dae6ab67ab82ea6b8e2 I get

[tdiesler@tdbell core]$ mvn -Dtest=MOD65TestCase install

testCircularityError(org.jboss.test.osgi.modules.MOD65TestCase)  Time elapsed: 0.616 sec  <<< ERROR!
java.lang.ClassNotFoundException: org/jboss/test/osgi/modules/a/CircularityError
at java.lang.Class.forName0(Native Method)
at java.lang.Class.forName(Class.java:247)
at org.jboss.test.osgi.modules.ModulesTestBase.loadClass(ModulesTestBase.java:129)
at org.jboss.test.osgi.modules.ModulesTestBase.assertLoadClass(ModulesTestBase.java:102)
at org.jboss.test.osgi.modules.MOD65TestCase.testCircularityError(MOD65TestCase.java:96)

with and without -Djboss.modules.lockless=true

**Comment by dmlloyd** [comment]
**[David Hildenbrand](https://issues.redhat.com/secure/ViewProfile.jspa?name=bosschaert)** commented

It would be nice to get this one fixed. It causes one of only 2 remaining OSGi Core TCK issues.

**Comment by dmlloyd** [comment]
@dmlloyd commented

Flavia looked at this in detail today.  Her findings were that the test is invalid because the class being loaded on lines 119/120 is actually being defined deeper in the stack trace, which can only ever result in ClassCircularityError.  If a class is being defined then it can't be loaded during that process, and there'd have to be a significant logic error there to explain that.  The reason you see the CNFE rather than the circularity error is because it is being caught at line 124 and wrapped in an IllegalStateException which gets discarded in favor of the CNFE.  If you preserve the error you'll find that it is propagated to the original calling code.

**Comment by dmlloyd** [comment]
@tdiesler commented

(02:13:49 PM) tdiesler: basically what we have in OSGi is the notion of lazy activation which gets triggered (among other things) by a class load from a bundle
(02:14:37 PM) tdiesler: The way this is currently implemented is by means of a hook into the LocalLoader like demonstrated by the test case
(02:15:14 PM) tdiesler: As dml points out, this is a hook into the class definition
(02:15:41 PM) tdiesler: and as it seems we cannot trigger another possibly circular class load from here
(02:15:42 PM) frainone: so, are you saying that a LocalLoader will attempt to load another class different than the one the LocalLoader was requested to load?
(02:15:54 PM) tdiesler: yes
(02:16:01 PM) tdiesler: it activates the bundle
(02:16:14 PM) tdiesler: the BundleActivator may load other classes
(02:16:22 PM) frainone: hm... yes, you cannot trigger it as you did in the test
(02:16:28 PM) tdiesler: and possibly the one that triggered it in a circular way
(02:16:39 PM) frainone: the JDK won't allow it :(
(02:17:16 PM) tdiesler: ok, we then need to find some other hook mechanism
(02:17:33 PM) tdiesler: into the the class load from a module
(02:18:12 PM) tdiesler: ideally some AfterClassLoadHook
(02:18:32 PM) tdiesler: perhaps you have an idea on how to do this with the current API
(02:18:53 PM) frainone: hm... I see. So, if you had a hook called after the class been loaded, it would fullfill your needs?
(02:20:05 PM) tdiesler: yes
(02:20:29 PM) frainone: from my point of view, it could be done. And your test would be the perfect start point to figure what pieces are needed
(02:21:18 PM) frainone: still, we need dmlloyd input on this
(02:22:25 PM) tdiesler: ok, have a look at ModuleManagerPluginImpl:280
(02:22:40 PM) tdiesler: in jbosgi-framework
(02:22:56 PM) tdiesler: for the whole story. There are filters involved too
(02:25:11 PM) tdiesler: Another good starting point is "4.4.6.2 Lazy Activation Policy" in the OSGi 4.2 Core spec

**Comment by dmlloyd** [comment]
@tdiesler commented

David, could you pls provide your feedback?

**Comment by dmlloyd** [comment]
@tdiesler commented

Raised to 'Critical' since this has been blocking OSGi Core 4.2 TCK compliance for a while

**Comment by dmlloyd** [comment]
@dmlloyd commented

Would a simple post-define-class hook suffice?  It could get the class that was defined, as well as the ClassSpec I guess?

**Comment by dmlloyd** [comment]
@dmlloyd commented

Please see http://github.com/dmlloyd/jboss-modules/tree/MODULES-65_postDefine for an implementation of this.

**Comment by dmlloyd** [comment]
@dmlloyd commented

Reassigning for comment.

**Comment by dmlloyd** [comment]
@tdiesler commented

Using this approach gives

```
java.lang.IllegalStateException: Cannot post define: interface org.jboss.test.osgi.modules.b.CircularityErrorDep
	at org.jboss.test.osgi.modules.MOD65TestCase$PostDefineModuleClassLoader.postDefine(MOD65TestCase.java:156)
	at org.jboss.modules.ModuleClassLoader.defineClass(ModuleClassLoader.java:384)
	at org.jboss.modules.ModuleClassLoader.loadClassLocal(ModuleClassLoader.java:252)
	at org.jboss.modules.ModuleClassLoader$1.loadClassLocal(ModuleClassLoader.java:75)
	at org.jboss.modules.Module.loadModuleClass(Module.java:578)
        ...
Caused by: java.lang.ClassCircularityError: org/jboss/test/osgi/modules/a/CircularityError
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClassCond(ClassLoader.java:632)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:616)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:141)
	at org.jboss.modules.ModuleClassLoader.defineClass(ModuleClassLoader.java:382)
	at org.jboss.modules.ModuleClassLoader.loadClassLocal(ModuleClassLoader.java:252)
	at org.jboss.modules.ModuleClassLoader$1.loadClassLocal(ModuleClassLoader.java:75)
	at org.jboss.modules.Module.loadModuleClass(Module.java:578)
	at org.jboss.modules.ModuleClassLoader.findClass(ModuleClassLoader.java:180)
	at org.jboss.modules.ConcurrentClassLoader.performLoadClassChecked(ConcurrentClassLoader.java:357)
	at org.jboss.modules.ConcurrentClassLoader.performLoadClass(ConcurrentClassLoader.java:306)
	at org.jboss.modules.ConcurrentClassLoader.loadClass(ConcurrentClassLoader.java:100)
	at org.jboss.test.osgi.modules.MOD65TestCase$PostDefineModuleClassLoader.postDefine(MOD65TestCase.java:153)
	... 51 more
```

https://github.com/tdiesler/jbosgi-framework/branches/JB[OSGI-442](https://issues.redhat.com/browse/OSGI-442)

The use case is that a class load from certain paths may trigger module activation (i.e. ModuleActivator is called).
During module activation a class gets loaded that has a dependency on the class which originally caused the module activation

Checking the spec whether the module can be activated asynchronously ...

**Comment by dmlloyd** [comment]
@tdiesler commented

> When a class load triggers the lazy activation, the Framework must first
> define the triggering class. This definition can trigger additional lazy activa-
> tions. These activations must be deferred until all transitive class loads and
> defines have finished. Thereafter, the activations must be executed in the
> reverse order of detection. That is, the last detected activation must be exe-
> cuted first. Only after all deferred activations are finished must the class
> load that triggered the activation return with the loaded class. If an error
> occurs during this process, it should be reported as a Framework ERROR
> event. However, the class load must succeed normally. A bundle that fails its
> lazy activation should not be activated again until the framework is
> restarted or the bundle is explicitly started by calling the Bundle start
> method.

I am taking another look to get this implemented at the osgi level.
The postDefine hook may be sufficient.

**Comment by dmlloyd** [comment]
@dmlloyd commented

Sounds like you could still do it with the hook by maintaining a thread-local stack (deque) of classes to consider for activation which you continually add to...

You'd just need some way to know what class originated the activation.  Some kind of invocation depth count or something?  I could possibly add that as a parameter if that seems like a plausible approach.

**Comment by dmlloyd** [comment]
@dmlloyd commented

I'll push the postDefine hook upstream for now and include it in the CR1 release.  If it doesn't work out we can change it before GA.

**Comment by dmlloyd** [comment]
@dmlloyd commented

I will optimistically call this resovled; re-open if it turns out I'm wrong.

**Comment by dmlloyd** [comment]
@tdiesler commented

OSGi expects reverse depth first order of detection

For a graph like this

```
TestChain
 +- AChain1
 |  +- AChain2
 |  +- CChain1
 |  +- BChain1
 +- BChain2
    +- CChain2
    +- DChain1
```

it expects an order of: T,A,C,B,D in reverse

I believe I need a hook that gets called in the order the nodes are visited not in the order the nodes are completed.

I'll look at the option of a preDefine hook, that in combination with what we already have should do the trick.

**Comment by dmlloyd** [comment]
@tdiesler commented

Done, please pull

https://github.com/tdiesler/jboss-modules/commit/a4cb6f0598d532e76aa49cd7b9cd008c397521fb

**Comment by dmlloyd** [comment]
@tdiesler commented

Please release at your earliest convenience. This would allow me to do our first OSGi 4.2 compliant framework release.
It's been a pleasure working with you.

**Comment by dmlloyd** [comment]
**[David Hildenbrand](https://issues.redhat.com/secure/ViewProfile.jspa?name=bosschaert)** commented

Yay - this is great progress!
